data_structures/src/transaction_factory.rs

```diff
@@ -121,23 +136,96 @@ pub trait OutputsCollection {
             )
             .ok_or_else(|| TransactionError::OutputValueOverflow)?;
 
-        let amount = output_value
-            .checked_add(fee)
-            .ok_or_else(|| TransactionError::OutputValueOverflow)?;
-
-        let (output_pointers, input_value) =
-            self.take_enough_utxos(amount, timestamp, block_number_limit, utxo_strategy)?;
-        let inputs: Vec<Input> = output_pointers.into_iter().map(Input::new).collect();
+        // For the first estimation: 1 input and 1 output more for the change address
+        let mut current_weight = calculate_weight(1, outputs.len() + 1, dr_output, max_weight)?;
+
+        let max_iterations = 1 + ((u64::from(max_weight) - current_weight) / u64::from(INPUT_SIZE));
+        for _i in 0..max_iterations {
+            let fee = match fee_type {
+                FeeType::Absolute => fee,
+                FeeType::Weighted => fee
+                    .checked_mul(current_weight)
+                    .ok_or_else(|| TransactionError::FeeOverflow)?,
+            };
+
+            let amount = output_value
+                .checked_add(fee)
+                .ok_or_else(|| TransactionError::FeeOverflow)?;
+
+            let (output_pointers, input_value) =
+                self.take_enough_utxos(amount, timestamp, block_number_limit, utxo_strategy)?;
+            let inputs: Vec<Input> = output_pointers.into_iter().map(Input::new).collect();
+
+            match fee_type {
+                FeeType::Absolute => {
+                    return Ok(TransactionInfo {
+                        inputs,
+                        outputs,
+                        input_value,
+                        output_value,
+                        fee,
+                    });
+                }
+                FeeType::Weighted => {
+                    let new_weight =
+                        calculate_weight(inputs.len(), outputs.len() + 1, dr_output, max_weight)?;
+                    if new_weight == current_weight {
+                        return Ok(TransactionInfo {
+                            inputs,
+                            outputs,
+                            input_value,
+                            output_value,
+                            fee,
+                        });
+                    } else {
+                        current_weight = new_weight;
+                    }
+                }
+            }
+        }
 
-        Ok(TransactionInfo {
-            inputs,
-            outputs,
-            input_value,
-            output_value,
-        })
+        unreachable!("Unexpected exit in build_inputs_outputs method");
     }
 }
 
+/// Calculate weight from inputs and outputs information
+pub fn calculate_weight(
+    inputs_count: usize,
+    outputs_count: usize,
+    dro: Option<&DataRequestOutput>,
+    max_weight: u32,
+) -> Result<u64, TransactionError> {
```

Comment:
We should always use the same units for weight, and I think the correct one is `u32`.

```suggestion
    max_weight: u32,
) -> Result<u32, TransactionError> {
```